ddl/ddl_api.go

```diff
@@ -2528,10 +2529,17 @@ func (d *ddl) DropColumn(ctx sessionctx.Context, ti ast.Ident, spec *ast.AlterTa
 }
 
 // modifiableCharsetAndCollation returns error when the charset or collation is not modifiable.
-func modifiableCharsetAndCollation(toCharset, toCollate, origCharset, origCollate string) error {
+// needRewriteCollationData is used when trying to modify the collation of a column, it is true when the column is with
+// index because index of a string column is collation-aware.
+func modifiableCharsetAndCollation(toCharset, toCollate, origCharset, origCollate string, needRewriteCollationData bool) error {
 	if !charset.ValidCharsetAndCollation(toCharset, toCollate) {
 		return ErrUnknownCharacterSet.GenWithStack("Unknown character set: '%s', collation: '%s'", toCharset, toCollate)
 	}
+
+	if needRewriteCollationData && collate.NewCollationEnabled() && !collate.CompatibleCollate(origCollate, toCollate) {
```

Comment:
`CompatibleCollate` is not needed here, since it will be checked later.

Reply:
Where will`CompatibleCollate ` be checked?

Reply:
L2543-L2557, I think that is more strict than `CompatibleCollate`

Reply:
Sound reasonable, maybe need to change the error in line  #2556

Reply:
No... Please notice that L2547 returned without more checks. For example, the following statements will be allowed:

```sql
# change from utf8mb4_bin to utf8mb4_general_ci
tidb> alter table t modify c varchar(10) collate utf8mb4_general_ci;
```